api/rest/restapi.go

```diff
@@ -362,60 +371,207 @@ func (api *API) graphHandler(w http.ResponseWriter, r *http.Request) {
 	sendResponse(w, err, graph)
 }
 
+func (api *API) consumeLocalAdd(args map[string]string, outObj *types.NodeSerial) error {
+	//TODO: when ipfs add starts supporting formats other than
+	// v0 (v1.cbor, v1.protobuf) we'll need to update this
+	b := types.BlockWithFormat{
+		Data:   outObj.Data,
+		Format: "",
+	}
+	args["cid"] = outObj.Cid // root node stored on last call
+	var hash string
+	err := api.rpcClient.Call("",
+		"Cluster",
+		"IPFSBlockPut",
+		b,
+		&hash)
+	if outObj.Cid != hash {
+		logger.Warningf("mismatch. node cid: %s\nrpc cid: %s", outObj.Cid, hash)
+	}
+	return err
+}
+
+func (api *API) finishLocalAdd(args map[string]string) error {
+	// TODO: pin the root node from here, rpc.Call(pin, args["cid"])
+	return nil
+}
+
+func (api *API) consumeShardAdd(args map[string]string, outObj *types.NodeSerial) error {
```

Comment:
I don't think so, though on a related not the consumeImport approach is maybe not go-ish or cluster-ish enough for your tastes.  

To avoid writing the same code twice I parametrized the code that reads from importer channels with `consume` and `finish` functions.  Subsequent calls need to communicate data to future calls, but the data they need to communicate is depended on the `consume` `finish` function implementation.  Specifically the ShardAdd functions need to communicate the shardID across calls and the `consumeLocalAdd` function should communicate the root Cid so that `finishLocalAdd` can pin the root of the file.  

I could change to a specific struct that has both shardID and rootCid fields but it doesn't seem to be a better fit than a map[string]string from my point of view.

Reply:
I see, it's not pretty. I might take a stab at that code and see if I can improve something